src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlRetry/SqlRetry.cs

```diff
@@ -0,0 +1,280 @@
+            }
+
+            if (sqlRetryServiceOptions.Value.AddTransientErrors != null)
+            {
+                transientErrors.UnionWith(sqlRetryServiceOptions.Value.AddTransientErrors);
+            }
+
+            maxRetries = sqlRetryServiceOptions.Value.MaxRetries;
+            retryMillisecondsDelay = sqlRetryServiceOptions.Value.RetryMillisecondsDelay;
+
+            EnsureArg.IsNotNull(sqlRetryServiceDelegateOptions, nameof(sqlRetryServiceDelegateOptions));
+            defaultIsExceptionRetriableOff = sqlRetryServiceDelegateOptions.DefaultIsExceptionRetriableOff;
+            customIsExceptionRetriable = sqlRetryServiceDelegateOptions.CustomIsExceptionRetriable;
+        }
+
+        public delegate bool IsExceptionRetriable(Exception ex);
+
+        public delegate Task ExecuteSqlWithRetriesAction(SqlCommand sqlCommand, CancellationToken cancellationToken);
+
+        public delegate Task<bool> ExecuteRowProcessor<T>(SqlDataReader reader, T result, CancellationToken cancellationToken);
+
+        public delegate T InitializeRowProcessor<T>();
+
+        public delegate void SqlConnectionInitializer(SqlConnection sqlConnection);
+
+        // TODO also define toT action delegate.
+
+        private static bool DefaultIsExceptionRetriable(Exception ex)
+        {
+            if (ex is SqlException sqlEx)
+            {
+                if (sqlEx.Number == 121 && sqlEx.Message.Contains("an error occurred during the pre-login handshake", StringComparison.OrdinalIgnoreCase))
+                {
+                    return true;
+                }
+
+                if (sqlEx.ToString().Contains("login failed", StringComparison.OrdinalIgnoreCase))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private bool RetryTest(Exception ex)
+        {
+            if (ex is SqlException sqlEx && transientErrors.Contains(sqlEx.Number))
+            {
+                return true;
+            }
+
+            if (!defaultIsExceptionRetriableOff && defaultIsExceptionRetriable(ex))
+            {
+                return true;
+            }
+
+            if (customIsExceptionRetriable(ex))
+            {
+                return true;
+            }
+
+            if (ex.IsRetryable())
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public async Task ExecuteSqlWithRetries(SqlCommand sqlCommand, SqlConnectionInitializer sqlConnectionInitializer, ExecuteSqlWithRetriesAction action, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNull(sqlCommand, nameof(sqlCommand));
+            EnsureArg.IsNotNull(action, nameof(action));
+
+            sqlCommand.CommandTimeout = (int)_sqlServerDataStoreConfiguration.CommandTimeout.TotalSeconds;
```

Comment:
>sqlCommand.CommandTimeout = (int)_sqlServerDataStoreConfiguration.CommandTimeout.TotalSeconds; [](http://example.com/codeflow?start=12&length=94)

We should either preserve value on input sql command or provide parameter to override configuration value.

Reply:
This comes from same place currently timeout comes from. So, we are now hardcoding timeouts into the source code?

Reply:
Each method that invokes particular SQL command has knowledge about data volume and, therefore, expectations on how long database call should last. Even the same SQL command invoked with different parameters values might yield substantially different results that might require different command timeouts. Only code knows about parameter values, so code should have ability to overwrite default configuration value. This could be simple "hardcoding" on per command basis or more advanced.

Reply:
OK, just to make it clear, so we are hardcoding timeouts like above "1200" into the source code?

Reply:
On the call/value level - yes.